docs/src/main/tut/freemonad.md

```diff
@@ -348,6 +348,120 @@ it's not too hard to get around.)
 val result: Map[String, Int] = compilePure(program, Map.empty)
 ```
 
+## Composing Free monads ADTs.
+
+Real world applications often time combine different algebras. 
+The `Inject` typeclass described by Swierstra in [Data types à la carte](http://www.staff.science.uu.nl/~swier004/Publications/DataTypesALaCarte.pdf)
+let us compose different algebras in the context of `Free`.
+
+Let's see a trivial example of unrelated ADT's getting composed as a `Coproduct` that conform a more complex program.
+
+```tut
+import cats.arrow.NaturalTransformation
+import cats.data.{Xor, Coproduct}
+import cats.free.{Inject, Free}
+import cats.{Id, ~>}
+import scala.collection.mutable.ListBuffer
+```
+
+```tut
+/* Handles user interaction */
+sealed trait Interact[A]
+case class Ask(prompt: String) extends Interact[String]
+case class Tell(msg: String) extends Interact[Unit]
+
+/* Represents persistence operations */
+sealed trait DataOp[A]
+case class AddCat(a: String) extends DataOp[Unit]
+case class GetAllCats() extends DataOp[List[String]]
+```
+
+Once the ADTs are defined we can formally state that a `Free` program is the Coproduct of it's Algebras.
+
+```tut
+type CatsApp[A] = Coproduct[DataOp, Interact, A]
+```
+
+In order to take advantage of monadic composition we use smart constructors to lift our Algebra to the `Free` context.
+
+```tut
+def lift[F[_], G[_], A](fa: F[A])(implicit I: Inject[F, G]): Free[G, A] =
+  Free.liftF(I.inj(fa))
```

Comment:
@ceedubs @fthomas Do you think we can use the `PartiallyApplied` trick given `Free.inject[App](Test1(x))` would not be able to infer `TestAlgebra1`? `Test1` is an `F[_]` but our `Coproduct` composition depends on `TestAlgebra1`. I can take it as far as inferring `A` in `Test1[A]` with that trick but not sure how would I go about auto inferring the parent sealed trait in the ADT. 
This would fail not finding a `Inject[X, Test1]` implicit when really the one we are supposed to get is `Inject[X, Test1Algebra]`.
```scala
final class FreeInjectPartiallyApplied[G[_]] private[free] {
    def apply[F[_], A](fa: F[A])(implicit I : Inject[F, G]): Free[G, A] =
      Free.liftF(I.inj(fa))
  }

  def inject[G[_]]: FreeInjectPartiallyApplied[G] = new FreeInjectPartiallyApplied
```